clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1256,65 +1182,331 @@ static void populateIntHeader(SYCLIntegrationHeader &H, const StringRef Name,
+        FuncType, Ctx.getTrivialTypeSourceInfo(Ctx.VoidTy), SC_None);
+    FD->setImplicitlyInline(IsInline);
+    setKernelImplicitAttrs(Ctx, FD, Name);
+
+    // Add kernel to translation unit to see it in AST-dump.
+    Ctx.getTranslationUnitDecl()->addDecl(FD);
+    return FD;
+  }
+
+public:
+  SyclKernelDeclCreator(Sema &S, SyclKernelFieldChecker &ArgChecker,
+                        StringRef Name, SourceLocation Loc, bool IsInline)
+      : SyclKernelDeclBase(
+            S, initKernelObj(S.getASTContext(), Name, Loc, IsInline), ArgTys,
+            Params),
+        ArgChecker(ArgChecker), FuncContext(SemaRef, KernelObj) {}
+
+  ~SyclKernelDeclCreator() {
+    KernelObj->setParams(Params);
+
+    if (ArgChecker.isValid())
+      SemaRef.addSyclDeviceDecl(KernelObj);
+  }
+
+  void handleSyclSamplerType(const FieldDecl *FD, QualType ArgTy) final {
+    handleSpecialType(FD, ArgTy);
+  }
+
+  void handlePointerType(const FieldDecl *FD, QualType ArgTy) final {
+    // TODO: Can we document what the heck this is doing?!
+    QualType PointeeTy = ArgTy->getPointeeType();
+    Qualifiers Quals = PointeeTy.getQualifiers();
+    Quals.setAddressSpace(LangAS::opencl_global);
+    PointeeTy = SemaRef.getASTContext().getQualifiedType(
+        PointeeTy.getUnqualifiedType(), Quals);
+    QualType ModTy = SemaRef.getASTContext().getPointerType(PointeeTy);
+    addParam(FD, ModTy);
+  }
+  void handleScalarType(const FieldDecl *FD, QualType ArgTy) final {
+    addParam(FD, ArgTy);
+  }
+
+  // This is implemented here because this is the only case where the recurse
+  // object is required.  The base type is pretty cheap, so we might opt
+  // to just always create it (the way this one is implemented) and just put
+  // this implementation in the base.
+  void handleStructType(const FieldDecl *FD, QualType ArgTy) final {
+    addParam(FD, ArgTy);
+
+    // Create descriptors for each accessor field in the class or struct
+    const auto *Wrapper = ArgTy->getAsCXXRecordDecl();
+    SyclKernelDeclBase Recurse(SemaRef, KernelObj, ArgTys, Params);
+    VisitKernelFields(Wrapper->fields(), Recurse);
+  }
+
+  void setBody(CompoundStmt *KB) { KernelObj->setBody(KB); }
+};
+
+class SyclKernelBodyCreator
+    : public SyclKernelFieldHandler<SyclKernelBodyCreator> {
+  SyclKernelDeclCreator &DeclCreator;
+  // TODO: When/Where does this get created?
+  CompoundStmt *KernelBody = nullptr;
+
+public:
+  SyclKernelBodyCreator(Sema &S, SyclKernelDeclCreator &DC)
+      : SyclKernelFieldHandler(S), DeclCreator(DC) {}
+  ~SyclKernelBodyCreator() { DeclCreator.setBody(KernelBody); }
+};
+
+class SyclKernelIntHeaderCreator
+    : public SyclKernelFieldHandler<SyclKernelIntHeaderCreator> {
+public:
+  SyclKernelIntHeaderCreator(Sema &S) : SyclKernelFieldHandler(S) {}
+};
+} // namespace
 
 // Generates the OpenCL kernel using KernelCallerFunc (kernel caller
-// function) defined is SYCL headers.
+// function) defined is Sycl headers.
```

Comment:
Why?

Reply:
I was being incredibly inconsistent with my naming at one point, so I ran a find/replace.  I guess it got a little out of hand :)